[MASWE-0009] Add Weak Cryptographic Key Generation (by appknox) (#2849)
OWASP/owasp-mastg: tests-beta/android/MASVS-CRYPTO/MASTG-TEST-0208.md

Reviewer: Please take a look at the new version of tests-beta/ios/MASVS-CRYPTO/MASTG-TEST-0209.md and update this one accordingly.

---
platform: android

> title: Weak Cryptographic Key Generation 

Reviewer 2: ```suggestion
title: Inappropriate Key Sizes 
```

id: MASTG-TEST-0208
type: [static]
weakness: MASWE-0009
---

## Overview

> The application appears to utilize a symmetric algorithm  with a weak key configuration, making it susceptible to multiple vulnerabilities and potential security breaches.

Reviewer 2: ```suggestion
In this test case, we will look for the use inappropriate key sizes in Android apps. To do this, we need to focus on the cryptographic frameworks and libraries that are available in Android and the methods that are used to generate cryptographic keys.

The application appears to utilize a symmetric algorithm  with a weak key configuration, making it susceptible to multiple vulnerabilities and potential security breaches.
```

> The application appears to utilize a symmetric algorithm  with a weak key configuration, making it susceptible to multiple vulnerabilities and potential security breaches.

Reviewer 2: ```suggestion
In this test case, we will look for the use inappropriate key sizes in Android apps. To do this, we need to focus on the cryptographic frameworks and libraries that are available in Android and the methods that are used to generate cryptographic keys.

In Android the class `KeyGenerator` offers different [encryption algorithms](https://docs.oracle.com/javase/8/docs/technotes/guides/security/StandardNames.html#KeyGenerator) for key creation. 

There are a variety of different [KeyGenerator.init()](https://developer.android.com/reference/javax/crypto/KeyGenerator#public-methods) methods to initialise the key generation, where some require random bytes. The entropy of these random bytes will rely on the  `SecureRandom` implementation in the available security provider, otherwise a system-provided source of randomness will be used.

```

## Steps

> 1. Run a static analysis tool such as @MASTG-TOOL-0110 on the code and look for uses of algorithms with insufficient key length.

Reviewer 2: ```suggestion
1. Run a static analysis tool such as @MASTG-TOOL-0110 on the code and look for uses of the cryptographic functions that generate keys.
```

## Observation

> The output should contain a **list of locations where insufficient key lengths are used** .

Reviewer 2: ```suggestion
The output should contain a **list of locations where insufficient key lengths are used**.
```

> The output should contain a **list of locations where insufficient key lengths are used**.

Reviewer 2: ```suggestion
The output should contain a list of locations where insufficient key lengths are used.
```

## Evaluation

> The test case fails if you can find the use of inappropriate key sizes within the source code.

Reviewer 2: ```suggestion
The test case fails if you can find the use of inappropriate key sizes within the source code. For example, a 1024-bit key size is considered weak for RSA encryption and a 128-bit key size is considered weak for AES encryption considering quantum computing attacks.
```

> The test case fails if you can find the use of algorithm using appropriate key size within the source code. This also fails if less popular algorithms are used.

Reviewer 2: ```suggestion
The test case fails if you can find the use of algorithms using inappropriate key sizes within the source code. This also fails if deprecated algorithms are used.
```

> The test case fails if you can find the use of algorithms using inappropriate key sizes within the source code. This also fails if deprecated algorithms are used.

Reviewer 2: ```suggestion
The test case fails if you can find the use of inappropriate key sizes within the source code. 
```
